pkg/process/process.go

```diff
@@ -0,0 +1,234 @@
+// Copyright 2022-2023 The Parca Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package process
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/prometheus/procfs"
+
+	"github.com/parca-dev/parca-agent/pkg/cgroup"
+)
+
+// key is a unique identifier for a process.
+// TODO(kakkoyun): This is not unique enough. We will need to add more fields.
+// For example, we can use the process start time to make sure we are not.
+// However, reading the stats file is expensive. For now, we will rely on the PID.
+// I will keep the struct to make it easier to add more fields in the future.
+type key struct {
+	pid int
+}
+
+type Tree struct {
+	tree        map[key]*process
+	mtx         *sync.RWMutex
+	disappeared map[key]struct{}
+	reused      map[key]*process
+}
+
+type process struct {
+	procfs.Proc
+
+	tree *Tree
+
+	parent int
+	// name of the field is the same as the one in kernel struct.
+	starttime uint64
+}
+
+func (p *process) ancestors() []*procfs.Proc {
+	p.tree.mtx.RLock()
```

Comment:
Since only the accesses need to be locked, let's move the locking into a `Get` function bound to the tree, and then just call `p.tree.Get(k)` in this function.

Reply:
Good call 👍